File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:

    bool isPalindrome(string s,int start, int end){
        while(start <= end){
            if(s[start]!= s[end]){
                return false;
            }
            start++;
            end--;
        }
        return true;
    }

    void recur(int ind, string s, vector<string>&path, vector<vector<string>>&res){
        if(ind == s.size()){
            res.push_back(path);
            return;
        }

        for(int i = ind; i < s.size(); i++){
            if(isPalindrome(s, ind, i)){
                path.push_back(s.substr(ind, i-ind+1));
                recur(i+1, s, path, res);
                path.pop_back();
            }
        }
    }

    vector<vector<string>> partition(string s) {
        vector<string> path;
        vector<vector<string>> res;
        recur(0, s, path, res);
        return res;
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def isValid(self, s, ind, i):
        return s[ind:i+1] == s[ind:i+1][::-1]

    def recur(self, ind, s, ans, path):
        if ind == len(s):
            ans.append(path[:])
            return 
        
        for i in range(ind, len(s)):
            if self.isValid(s, ind, i):
                path.append(s[ind:i+1])
                self.recur(i + 1, s, ans, path)
                path.pop()


    def partition(self, s: str) -> List[List[str]]:
        ans = []
        path = []
        n = len(s)
        self.recur(0, s, ans, path)
        return ans